src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -53,64 +53,221 @@ static ucs_status_t uct_rdmacm_cm_query(uct_cm_h cm, uct_cm_attr_t *cm_attr)
-    priv_data_ret = cep->wireup.priv_pack_cb(cep->user_data, dev_name, hdr + 1);
+    ucs_free((void*)conn_param.private_data);
+}
 
-    if ((priv_data_ret < 0) || (priv_data_ret > uct_rdmacm_cm_get_max_conn_priv())) {
-        ucs_error("rdma_cm private data pack function on client ep %p returned "
-                  "%zd (max: %zu)", cep, priv_data_ret,
-                  uct_rdmacm_cm_get_max_conn_priv());
-        return;
+static ucs_status_t uct_rdmacm_cm_id_to_dev_addr(struct rdma_cm_id *cm_id,
+                                                 uct_device_addr_t **dev_addr_p,
+                                                 size_t *dev_addr_len_p)
+{
+    struct ibv_port_attr port_arrt;
+    uct_ib_address_t *dev_addr;
+    struct ibv_qp_attr qp_attr;
+    size_t addr_length;
+    int qp_attr_mask;
+
+    qp_attr.qp_state = IBV_QPS_RTR;
+    /* get the qp attributes in order to modify the qp state.
+     * the ah_attr fields from them are required to extract the device address
+     * of the remote peer.
+     */
+    if (rdma_init_qp_attr(cm_id, &qp_attr, &qp_attr_mask)) {
+        ucs_error("rdma_init_qp_attr failed. id %p, state: %d. %m",
+                  cm_id, qp_attr.qp_state);
+        return UCS_ERR_IO_ERROR;
+     }
+
+    if (ibv_query_port(cm_id->verbs, cm_id->port_num, &port_arrt)) {
+        ucs_error("ibv_query_port failed. %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    addr_length = uct_ib_address_size(&qp_attr.ah_attr.grh.dgid, 0,
+                                      port_arrt.link_layer == IBV_LINK_LAYER_ETHERNET);
+
+    dev_addr = ucs_malloc(addr_length, "IB device address");
+    if (dev_addr == NULL) {
+        return UCS_ERR_NO_MEMORY;
     }
 
-    ucs_assert(priv_data_ret <= UINT8_MAX);
-    hdr->length = (uint8_t)priv_data_ret;
-    hdr->status = UCS_OK;
+    uct_ib_address_pack(&qp_attr.ah_attr.grh.dgid, qp_attr.ah_attr.dlid,
+                        port_arrt.link_layer == IBV_LINK_LAYER_ETHERNET, 0,
+                        dev_addr);
+
+    *dev_addr_p     = (uct_device_addr_t *)dev_addr;
+    *dev_addr_len_p = addr_length;
+    return UCS_OK;
+}
+
 
-    status = uct_rdamcm_cm_ep_set_qp_num(&conn_param, hdr, cep);
+static void uct_rdmacm_cm_handle_event_connect_request(struct rdma_cm_event *event)
+{
+    uct_rdmacm_priv_data_hdr_t *hdr;
+    uct_rdmacm_listener_t      *listener;
+    char                       dev_name[UCT_DEVICE_NAME_MAX];
+    uct_device_addr_t          *dev_addr;
+    size_t                     addr_length;
+    uct_cm_remote_data_t       remote_data;
+    ucs_status_t               status;
+
+    listener = event->listen_id->context;
+    hdr      = (uct_rdmacm_priv_data_hdr_t *)event->param.conn.private_data;
+    ucs_assert(hdr->status == UCS_OK);
+
+    uct_rdmacm_cm_id_to_dev_name(event->id, dev_name);
+
+    status = uct_rdmacm_cm_id_to_dev_addr(event->id, &dev_addr, &addr_length);
     if (status != UCS_OK) {
+        hdr->status = UCS_ERR_REJECTED;
```

Comment:
`hdr->status = status` would notify the client on rdmacm_cm level the reason of reject